subprojects/plugins/src/main/java/org/gradle/jvm/component/internal/JvmSoftwareComponentInternal.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.component.internal;
+
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.component.SoftwareComponent;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.SourceSetOutput;
+import org.gradle.api.tasks.TaskProvider;
+import org.gradle.api.tasks.bundling.Jar;
+import org.gradle.api.tasks.compile.JavaCompile;
+
+/**
+ * A {@link SoftwareComponent} which produces variants intended for use within the JVM ecosystem.
+ * <p>
+ * There is currently no public interface for this type of component, as the JVM component
+ * infrastructure is still under construction. The main blocker for publicizing this component
+ * is the lack of multi-target functionality. This interface currently exposes the
+ * {@link #getJar()}, {@link #getSources()}, {@link #getImplementation()}, etc. methods, which
+ * imply that this component only supports a single compilation target. A multi-target component
+ * would not expose these methods at the top-level, but would instead encapsulate them within
+ * some kind of {@code targets} or {@code pipelines} container.
+ * <p>
+ * Before publicizing this component, we also need to consider how component extensibility works.
+ * For example, for the java-library plugin we have the additional {@code api} and {@code compileOnlyApi}
+ * scopes. Do we expect that plugin to write a new component interface and extend this in order to add the
+ * proper getter methods? What about the concrete component class which implements that new interface? Does
+ * it extend the default implementation class? Is there a way we can avoid Java inheritance?
+ */
+public interface JvmSoftwareComponentInternal extends SoftwareComponent {
+
+    /**
+     * Configures this component to publish a javadoc jar alongside the primary artifacts. As a result,
+     * this method also configures the necessary configurations and tasks required to produce
+     * the javadoc artifact.
+     */
+    void withJavadocJar();
```

Comment:
🤔 How about `publishJavadoc` and `publishSources` for these new methods?

They are not withers, the Jar suffix hides the fact that the metadata is also affected, and if I were a newcomer looking to publish javadoc I would start by looking at methods named `publishXYZ`.

Reply:
I want to avoid the `publish` terminology since these add variants and do not strictly publish these jars. But I agree these are not withers. Updated to `addJavadocJarVariant` and `addSourcesJarVariant`.